chapter04_convolutional-neural-networks/kaggle-gluon-cifar10.md

```diff
@@ -0,0 +1,412 @@
+
+* [训练数据集train.7z下载地址](https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/kaggle-cifar10/train.7z)
+* [测试数据集test.7z下载地址](https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/kaggle-cifar10/test.7z)
+* [训练数据标签trainLabels.csv下载地址](https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/kaggle-cifar10/trainLabels.csv)
+
+
+### 解压数据集
+
+训练数据集train.7z和测试数据集test.7z都是压缩格式，下载后请解压缩。假设解压缩后原始数据集的路径如下（data_dir是存放数据的路径，例如'..
+/data/kaggle_cifar10'，'train_dir'和'test_dir'分别是对train.7z和test.7z解压后的文件夹名称，例如'train'和'test'）：
+
+* data_dir/train_dir/1.png
+* data_dir/train_dir/...
+* data_dir/train_dir/50000.png
+* data_dir/test_dir/1.png
+* data_dir/test_dir/...
+* data_dir/test_dir/300000.png
+* data_dir/trainLabels.csv
+
+为了使网页编译快一点，我们在git repo里仅仅存放100个训练样本（'train_tiny.zip'）和1个测试样本（'test_tiny.zip'）。执行以下代码会从git repo里解压生成小样本训练和测试数据，文件夹名称分别为'train_tiny'和'test_tiny'。训练数据标签的压缩文件将被解压成trainLabels.csv。
+
+```{.python .input  n=1}
+# 本段代码仅为网页编译方便，无需本地执行。
+import zipfile
+for fin in ['train_tiny.zip', 'test_tiny.zip', 'trainLabels.csv.zip']:
+    with zipfile.ZipFile('../data/kaggle_cifar10/' + fin, 'r') as zin:
+        zin.extractall('../data/kaggle_cifar10/')
+```
+
+### 整理数据集
+
+我们定义下面的reorg_cifar10_data函数来整理数据集。整理后，同一类图片将出现在在同一个文件夹下，便于`Gluon`稍后读取。
+
+函数中的参数如data_dir、train_dir和test_dir对应上述数据存放路径及训练和测试的图片集文件夹名称。参数label_file为训练数据标签的
+文件名称。参数input_dir时整理后数据集文件夹名称。参数valid_ratio是验证集占原始训练集的比重。以valid_ratio=0.1为例，由于原始训
+练数据有5万张图片，调参时将有4万5千张图片用于训练（整理后存放在input_dir/train）而另外5千张图片为验证集（整理后存放在input_dir/va
+lid）。
+
+```{.python .input  n=1}
+import os
+import shutil
+
+def reorg_cifar10_data(data_dir, label_file, train_dir, test_dir, input_dir, valid_ratio):
+    idx_label = dict()
+    labels = set()
+    # 读取训练数据标签。
+    with open(os.path.join(data_dir, label_file), 'r') as label_f:
+        # 跳过文件头行（栏名称）。
+        next(label_f)
+        for line in label_f:
+            idx, label = line.rstrip().split(',')
+            idx_label[int(idx)] = label
+            labels.add(label)
+
+    num_train = len(os.listdir(os.path.join(data_dir, train_dir)))
+    num_train_tuning = int(num_train * (1 - valid_ratio))
+    assert 0 < num_train_tuning < num_train
+    num_train_tuning_per_label = num_train_tuning // len(labels)
+    label_count = dict()
+    
+    def mkdir_if_not_exist(path):
+        if not os.path.exists(os.path.join(*path)):
+            os.makedirs(os.path.join(*path))
+
+    # 整理训练和验证集。
+    for train_file in os.listdir(os.path.join(data_dir, train_dir)):
+        idx = int(train_file.split('.')[0])
+        label = idx_label[idx]
+        mkdir_if_not_exist([data_dir, input_dir, 'train_valid', label])
+        shutil.copy(os.path.join(data_dir, train_dir, train_file), 
+                    os.path.join(data_dir, input_dir, 'train_valid', label))
+        if label not in label_count or label_count[label] < num_train_tuning_per_label:
+            mkdir_if_not_exist([data_dir, input_dir, 'train', label])
```

Comment:
avoid make_dir and copy if possible

Reply:
it's required by vision.ImageFolderDataset